crates/metaslang/bindings/src/lib.rs

```diff
@@ -107,72 +137,103 @@ impl<KT: KindTypes + 'static> Bindings<KT> {
     }
 }
 
-pub struct Handle<'a, KT: KindTypes + 'static> {
+#[derive(Clone)]
+pub struct Definition<'a, KT: KindTypes + 'static> {
     owner: &'a Bindings<KT>,
     handle: GraphHandle,
 }
 
-impl<KT: KindTypes + 'static> Handle<'_, KT> {
-    pub fn is_definition(&self) -> bool {
-        self.owner.stack_graph[self.handle].is_definition()
+impl<'a, KT: KindTypes + 'static> Definition<'a, KT> {
+    pub fn get_cursor(&self) -> Option<Cursor<KT>> {
+        self.owner.cursors.get(&self.handle).cloned()
+    }
+
+    pub fn get_definiens_cursor(&self) -> Option<Cursor<KT>> {
+        self.owner.definiens.get(&self.handle).cloned()
+    }
+
+    pub fn get_file(&self) -> Option<&'a str> {
+        self.owner.stack_graph[self.handle]
+            .file()
+            .map(|file| self.owner.stack_graph[file].name())
+    }
+}
+
+impl<KT: KindTypes + 'static> Debug for Definition<'_, KT> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("BindingsHandle").field(&self.handle).finish()
     }
+}
 
-    pub fn is_reference(&self) -> bool {
-        self.owner.stack_graph[self.handle].is_reference()
+impl<KT: KindTypes + 'static> PartialEq for Definition<'_, KT> {
+    fn eq(&self, other: &Self) -> bool {
+        let our_owner: *const Bindings<KT> = self.owner;
+        let other_owner: *const Bindings<KT> = other.owner;
+        our_owner == other_owner && self.handle == other.handle
     }
+}
+
+#[derive(Clone)]
+pub struct Reference<'a, KT: KindTypes + 'static> {
+    owner: &'a Bindings<KT>,
+    handle: GraphHandle,
+}
 
+impl<'a, KT: KindTypes + 'static> Reference<'a, KT> {
     pub fn get_cursor(&self) -> Option<Cursor<KT>> {
         self.owner.cursors.get(&self.handle).cloned()
     }
 
-    pub fn get_file(&self) -> Option<&str> {
+    pub fn get_file(&self) -> Option<&'a str> {
```

Comment:
We earlier discussed storing the source file path (or "ID") on the cursor itself, so that users can just get it from cursors that we return from here.
I wonder if we still need that, or if this new API replaces that use case.

Reply:
This works only for the bindings handles, and this information comes from the stack graph. I think it would be better to get this from the CST cursor, but in the end both values should match.